src/main/scala/com/wavesplatform/network/NetworkServer.scala

```diff
@@ -0,0 +1,223 @@
+        new OptimisticExtensionLoader,
+        utxPoolSynchronizer,
+        scoreObserver,
+        coordinator -> coordinatorExecutor)))
+      .bind(bindAddress)
+      .channel()
+  }
+
+  private val outgoingChannelCount = new AtomicInteger(0)
+  private val channels = new ConcurrentHashMap[InetSocketAddress, Channel]
+
+  private val clientHandshakeHandler =
+    new HandshakeHandler.Client(handshake, peerInfo, peerUniqueness, blacklist)
+
+  private val bootstrap = new Bootstrap()
+    .group(workerGroup)
+    .channel(classOf[NioSocketChannel])
+    .handler(new PipelineInitializer[SocketChannel](Seq(
+      errorHandler,
+      new HandshakeDecoder,
+      new HandshakeTimeoutHandler,
+      clientHandshakeHandler,
+      lengthFieldPrepender,
+      new LengthFieldBasedFrameDecoder(1024*1024, 0, 4, 0, 4),
+      new LegacyFrameCodec,
+      discardingHandler,
+      messageCodec,
+      peerSynchronizer,
+      historyReplier,
+      new ExtensionSignaturesLoader(settings.synchronizationSettings.synchronizationTimeout, blacklist),
+      new ExtensionBlocksLoader(history, settings.synchronizationSettings.synchronizationTimeout, blacklist),
+      new OptimisticExtensionLoader,
+      utxPoolSynchronizer,
+      scoreObserver,
+      coordinator -> coordinatorExecutor)))
+
+  workerGroup.scheduleWithFixedDelay(1.second, 5.seconds) {
+    if (outgoingChannelCount.get() < settings.networkSettings.maxOutboundConnections) {
+      peerDatabase
+        .getRandomPeer(settings.networkSettings.declaredAddress.toSet ++ channels.keySet().asScala)
+        .foreach(connect)
+    }
+  }
+
+  def connect(remoteAddress: InetSocketAddress): Unit =
+    channels.computeIfAbsent(remoteAddress, _ => {
+      bootstrap.connect(remoteAddress)
+        .addListener { (connFuture: ChannelFuture) =>
+          if (connFuture.isDone) {
+            if (connFuture.cause() != null) {
+              log.debug(s"${id(connFuture.channel())} Connection failed, blacklisting $remoteAddress", connFuture.cause())
+              blacklist(remoteAddress.getAddress)
+            } else if (connFuture.isSuccess) {
+              log.debug(s"${id(connFuture.channel())} Connection established")
+              outgoingChannelCount.incrementAndGet()
+              connFuture.channel().closeFuture().addListener { (closeFuture: ChannelFuture) =>
+                val remainingCount = outgoingChannelCount.decrementAndGet()
+                log.debug(s"${id(closeFuture.channel)} Connection closed, $remainingCount outgoing channel(s) remaining")
+                allChannels.remove(closeFuture.channel())
+                channels.remove(remoteAddress, closeFuture.channel())
+              }
+              allChannels.add(connFuture.channel())
+            }
+          }
+        }.channel()
+    })
+
+  def writeToLocalChannel(message: AnyRef): Unit = localClientChannel.writeAndFlush(message)
```

Comment:
In order to communicate with itself, the application uses netty. This can't be a good sign. 
Probably we can extract some other layer of abstraction.